### DISCUSSION TITLE
Launching Tutorials!

### DISCUSSION THREAD

**Discussion by alexvoss** [squidfunk/mkdocs-material#7220] [Announcement]
Hi Community,

today we are launching a new element in our documentation. To make it easier for people to get started with some of the awesome features of Material for MkDocs, we have started to create [tutorials]. The first ones cover [blogs] and [social cards]. They not only guide you through setting up and configuring the blog- and social plugin but also aim to provide more background than the documentation provided so far.

We have decided to split the topics into smaller chunks so that each part should take no more than 30mins to complete. For each topic there is also a template repository that contains a project similar to what you should get after working through the corresponding tutorials. You can use this as a starting point for your own projects.

While we have tried hard to come up with content to meet the community's needs and have carefully reviewed the material, we would love to get feedback on what could be improved. Please just post in this thread if you think we got something wrong or if there are needs we have not addressed that you think many people have.

We will try to expand the list of available tutorials over time but this does compete with other work on the project, so it may take a while.

Best wishes,

@alexvoss and @squidfunk

[tutorials]: https://squidfunk.github.io/mkdocs-material/tutorials/
[blogs]: https://squidfunk.github.io/mkdocs-material/tutorials/#blogs
[social cards]: https://squidfunk.github.io/mkdocs-material/tutorials/#social-cards

**Comment by kamilkrzyskow** [comment]
Hi @alexvoss,
given that the provided templates are supposed to be a "working starting point" I noticed they lack GitHub Action workflows, and I wanted to ask if there was some sort of limitation to adding the workflow files to the templates, other than the differences between Insiders / Community version?

Using the workflow from the [Publishing Guide](https://squidfunk.github.io/mkdocs-material/publishing-your-site/) isn't enough to pass the Blog tutorial section, and I've noticed that a couple of weeks ago while helping out in a discussion:
- https://github.com/squidfunk/mkdocs-material/discussions/7336#discussioncomment-9991702
- https://github.com/kamilkrzyskow/test-insider-token/commits/main/
  To get it working I had to:
  - extend the template with the Insiders workflow
  - add `sudo` to install pngquant
  - export the `PYTHONPATH=.` environmental variable for Python to detect the `ext` module
  - install the rss-plugin

I wanted to first read the tutorials over to check if there are mentions of those actions, but I haven't gotten around to it. Just did a quick Ctrl + F and couldn't find some of them.

I got reminded about this because of another discussion:
- https://github.com/squidfunk/mkdocs-material/discussions/7485
where the user expected a more plug and play solution for serving the MkDocs page from the repository files. 

So I guess this is a topic worth looking into? I haven't checked the tutorials about Social cards.

**Comment by alexvoss** [comment]
Hi @kamilkrzyskow, 

thanks for the feedback. 

I guess there is no mention of the GitHub Actions because I considered deployment an orthogonal topic. Deciding how and where to deploy is not a trivial matter as [recent discussions about GitHub and Google Search](https://github.com/squidfunk/mkdocs-material/discussions/7478) have shown. I imagined people building locally and then taking things in whatever direction they choose. 

Given that deploying with Insiders requires fiddling with the `GH_TOKEN` and this can be a stumbling block, I think that it might be better to leave this out of the tutorial for fear of making it too complex. Perhaps a separate tutorial on deployment might be better, which could then be written to support all of the tutorials?

As to the specific issues:

- "add sudo to install pngquant " -  yes, that would be necessary, do you want to PR or should I? It is in the general deployment guide and is not part of the tutorial.
- "export the PYTHONPATH=." - this is needed for the custom slugify function and should be mentioned in that section. We had a discussion a while back about ways for avoiding this such as putting the code in a hooks directory instead, which MkDocs adds to the path automagically. Is it uglier than the need for to set the `PYTHONPATH`? Not sure. Any thoughts? 
- "install the rss-plugin" - that is needed for the Blog tutorial but the step to do it is, in fact, mentioned.

The issue with the minify plugin arose because the OP wanted to build the Material for MkDocs documentation. I am not sure we should set GH Actions up to add all that is required for that. Each documentation project will have slightly different requirements, after all.

Thanks for checking this.

**Reply by squidfunk** [reply]
I'm also on the fence whether deployment should be part of the tutorials, or whether we should just link to the [publishing guide](https://squidfunk.github.io/mkdocs-material/publishing-your-site/). The problem is the possibilities for deployments are so versatile, that it's quite hard to decide what we want to cover in a tutorial. GitHub Pages? S3? If we'd link to the publishing guide, we'd make it easier for users to choose among [different options](https://squidfunk.github.io/mkdocs-material/publishing-your-site/#other). Additionally, we consider GitHub Pages to be the canonical case for deployment, as it's free, which is also covered in the publishing guide.

However, we might add a GitHub Actions workflow to the tutorial, which just publishes to GitHub Pages, and explains how to set it up. This is also covered in the [YouTube video](https://www.youtube.com/watch?v=Q-YA_dA8C20&t=740s), which we link from our getting started page.

**Reply by kamilkrzyskow** [reply]
I would suggest to keep it as simple as possible 😄 As the Publishing Guide does provide up-to-date workflow files for GitHub Pages, I think it's reasonable to simply add an explicit section, inside the tutorials, about the needed modifications required to make in those files, to remove the need to look for that information. 

I'm just trying to lower the bar for entry to allow for a quicker setup, in general. Even if people don't have any clue what CI pipelines are, and how a workflow file works, they will be able to just put the files in the correct directory, make adjustments, and be done with it. 2-3 bullet points for steps in the tutorial.

> The problem is the possibilities for deployments are so versatile

Let's just focus on GitHub Pages, because if somebody goes the less mainstream path and would prefer to use GitLab CI or Woodpecker CI, then they should be prepared for more hurdles (aka first learning how GitHub Pages work and then rewriting it to their CI solution by hand or with AI LLMs) 

> "install the rss-plugin" - that is needed for the Blog tutorial but the step to do it is, in fact, mentioned.

Yes, it's mentioned, both in the template description and as a step in the Tutorial. However. some people just don't know where the installation step is, in the workflow file, because they never bothered to read it 🙂, but they would read it if the tutorial would point it out as "knowledge worth having" 😅 

> yes, that would be necessary, do you want to PR or should I? It is in the general deployment guide and is not part of the tutorial.

True, I will create the PR over the weekend, as I had some plans to modify the cache approach as well, to opt out of the time based solution, so this will be a general revamp/fixup of the CI workflows.

> We had a discussion a while back about ways for avoiding this such as putting the code in a hooks directory instead, which MkDocs adds to the path automagically. Is it uglier than the need for to set the PYTHONPATH?

I'm not sure if I remember that discussion, sorry. AFAIK, the hooks directory gets added to the `sys.path` only for the duration of their import, so that they can use other relative modules, and then the path gets removed to avoid module name conflicts. 
So I don't think the user will be able to add their `ext` module in the `mkdocs.yml` as the `sys.path` won't be able to find it, unless simply by importing the module it gets cached 🤔 If this approach will work, then it helps with the quick setup, as the template could just enable the hook.

A hook could also work via the `on_config` event and modify the `mdx_config` to inject our changes to the Markdown extension configuration, but I feel like this approach is a bit too "complex" compared to simply setting the variable. That's more in the realm of hacky solutions than easy getting started templates 😆 

I'll open up the PR with fixed workflow file examples, as for the matter how to include it in the tutorials, I'll leave it up to you.

**Reply by kamilkrzyskow** [reply]
> I will create the PR over the weekend

Yeah, this didn't work out, I didn't forget about it, and will have it done in the coming days :v:
